src/nerdbank-streams/src/Channel.ts

```diff
@@ -0,0 +1,135 @@
+import { Duplex } from "stream";
+import { ChannelOptions } from "./ChannelOptions";
+import { ControlCode } from "./ControlCode";
+import { Deferred } from "./Deferred";
+import { FrameHeader } from "./FrameHeader";
+import { IDisposableObservable } from "./IDisposableObservable";
+import { MultiplexingStreamClass } from "./MultiplexingStream";
+
+export abstract class Channel implements IDisposableObservable {
+    public duplex: Duplex;
+    public acceptance: Promise<void>;
+    public completion: Promise<void>;
+    private _isDisposed: boolean = false;
+
+    get isDisposed(): boolean {
+        return this._isDisposed;
+    }
+
+    public dispose() {
+        this._isDisposed = true;
+    }
+}
+
+// tslint:disable-next-line:max-classes-per-file
+export class ChannelClass extends Channel {
+    public readonly id: number;
+    private _duplex: Duplex;
+    private readonly _acceptance = new Deferred<void>();
+    private readonly _completion = new Deferred<void>();
+
+    constructor(
+        private multiplexingStream: MultiplexingStreamClass,
```

Comment:
style-wise, I'm not a fan of mixing these - my view is that they should either all be defined as constructor params or as fields, not mixed

Reply:
I just learned that I can actually add more modifiers (e.g. `readonly`) to these constructor params, which makes them more useful.
**But** I also learned that the doc comments I declare on the constructor parameters do _not_ propagate to the fields that the typescript compiler emits for me, so that's a real bummer that will prevent me from using the feature.